contractsV5/eth2dai/Eth2DaiReserve.sol

```diff
@@ -0,0 +1,871 @@
+            order0Pay = ask.payAmount;
+            // sell eth rate        : ask.buyAmount / ask.payAmount
+            // buy eth rate         : bid.payAmount / bid.buyAmount
+            // median rate (eth/dai): (ask.buyAmount / ask.payAmount + bid.payAmount / bid.buyAmount) / 2;
+            // take amt dai         : ask.payAmount
+            // -> amt eth           : ask.payAmount * (ask.buyAmount / ask.payAmount + bid.payAmount / bid.buyAmount) / 2;
+            order0Buy = (ask.buyAmount + ask.payAmount * bid.payAmount / bid.buyAmount) / 2;
+        }
+
+        uint ethOrderSize = isEthToToken ? srcAmount : srcAmount * order0Buy / order0Pay;
+
+        BasicDataConfig memory basicData = getTokenBasicData(token);
+
+        if (basicData.minETHSupport > ethOrderSize) {
+            maxTakes = 0;
+            maxTraverse = 0;
+            minPayAmount = 0;
+            return (maxTakes, maxTraverse, minPayAmount);
+        }
+
+        FactorDataConfig memory factorData = getFactorData(token);
+
+        maxTraverse = (factorData.maxTraverseX * ethOrderSize / PRECISION + factorData.maxTraverseY) / BASIC_FACTOR_STEP;
+        maxTraverse = minOf(maxTraverse, basicData.maxTraverse);
+
+        maxTakes = (factorData.maxTakeX * ethOrderSize / PRECISION + factorData.maxTakeY) / BASIC_FACTOR_STEP;
+        maxTakes = minOf(maxTakes, basicData.maxTakes);
+
+        uint minETHAmount = (factorData.minOrderSizeX * ethOrderSize + factorData.minOrderSizeY * PRECISION) / BASIC_FACTOR_STEP;
+
+        // translate min amount to pay token
+        minPayAmount = isEthToToken ? minETHAmount : minETHAmount * order0Pay / order0Buy;
+    }
+
+    function getFirstOffer(ERC20 offerSellGem, ERC20 offerBuyGem)
+        internal view
+        returns(uint offerId, uint offerPayAmount, uint offerBuyAmount)
+    {
+        offerId = otc.getBestOffer(offerSellGem, offerBuyGem);
+        (offerBuyAmount, ,offerPayAmount, ) = otc.getOffer(offerId);
+    }
+
+    function getNextBestOffer(
+        ERC20 offerSellGem,
+        ERC20 offerBuyGem,
+        uint payAmount,
+        uint prevOfferId)
+        public view
+        returns(
+            uint offerId,
+            uint offerPayAmount,
+            uint offerBuyAmount
+        )
+    {
+        if (prevOfferId == INVALID_ID) {
+            offerId = otc.getBestOffer(offerSellGem, offerBuyGem);
+        } else {
+            offerId = otc.getWorseOffer(prevOfferId);
+        }
+
+        (offerBuyAmount, ,offerPayAmount, ) = otc.getOffer(offerId);
+
+        while (payAmount > offerPayAmount) {
+            offerId = otc.getWorseOffer(offerId); // next best offer
+            if (offerId == 0) {
+                offerId = 0;
+                offerPayAmount = 0;
+                offerBuyAmount = 0;
+                break;
+            }
+            (offerBuyAmount, ,offerPayAmount, ) = otc.getOffer(offerId);
+        }
+    }
+
+    function getFirstBidAndAskOrdersPub(ERC20 token)
```

Comment:
do we need this here?
is it only for debugging?
is it relevant for debugging after we deploy?

Reply:
I removed this function, just for testing